src/common/tse.c

```diff
@@ -1032,6 +1032,61 @@ tse_task_reinit(tse_task_t *task)
 	return rc;
 }
 
+int
+tse_task_reset(tse_task_t *task, tse_task_func_t task_func, void *priv)
+{
+	struct tse_task_private		*dtp = tse_task2priv(task);
+	tse_sched_t			*sched = tse_task2sched(task);
+	struct tse_sched_private	*dsp = tse_sched2priv(sched);
+	int				rc;
+
+	D_CASSERT(sizeof(task->dt_private) >= sizeof(*dtp));
+
+	D_MUTEX_LOCK(&dsp->dsp_lock);
+
+	if (dsp->dsp_cancelling) {
+		D_ERROR("Scheduler is canceling, can't reset task\n");
+		D_GOTO(err_unlock, rc = -DER_NO_PERM);
+	}
+
+	if (dtp->dtp_running || dtp->dtp_completing) {
+		D_ERROR("Can't reset a task that is running.\n");
+		D_GOTO(err_unlock, rc = -DER_NO_PERM);
+	}
+
+	/** Mark the task back at init state */
+	dtp->dtp_running = 0;
+	dtp->dtp_completing = 0;
+	dtp->dtp_completed = 0;
+
+	/** reset stack pointer as zero */
+	if (dtp->dtp_stack_top != 0) {
+		D_ERROR("task %p, dtp_stack_top reset from %d to zero.\n",
+			task, dtp->dtp_stack_top);
+		dtp->dtp_stack_top = 0;
+	}
+
+	D_INIT_LIST_HEAD(&dtp->dtp_list);
```

Comment:
Seems should use d_list_del_init(). For example if that task was tse_task_schedule() with instant==false, it will be put in dsp->dsp_init_list and without setting dtp_running flag. in that case you need to decref dsp's refcount also.

I am not sure if there is some other issues for below list reinit.

To simplify the issue, maybe can only allow to reset a task in COMPLETED status?

Reply:
ok i agree that can simplify things. will update

Reply:
or we should just check and return error if they are not empty? I think it's safer if these link chains are list_empty()

Reply:
sure, that makes sense. i added those checks now